README.md
# Fast Scroll

[Fast Scroll](https://github.com/fuweichin/fast-scroll) by Fuwei Chin brings fast vertical scroll, fast horizontal scroll and fast zoom.



## Features

+ support both vertical and horizontal scroll directions, both instant and smooth scroll behaviors
+ support both by-pixels and by-page scroll modes
+ supports fast zoom iteration over integral multiples(200%, 300%, 400%, ...), unit fractions(50%, 33%, 25%, ...)
+ works friendly with overlay scrollbars if scroll behavior is smooth

Special Notes:

+ won't work when [Smooth Scroll Wheel](https://learn.microsoft.com/en-us/windows-hardware/drivers/hid/keyboard-and-mouse-hid-client-drivers#supported-buttons-and-wheels-on-mice) or or "Scroll one screen at a time" is detected, thus to avoid conflict
+ only works on top-level scrolling elements

Alternatives:

+ [Fast Scroll](https://chrome.google.com/webstore/detail/fast-scroll/ecnjcglleblahonnenpaiofkabfakgdi?hl=en-US) by Mickaël Allonneau
+ [Fast Scroll](https://chrome.google.com/webstore/detail/fast-scroll/dkdnncjokeklapahlhbgfnnakjjaogmb?hl=en-US) by Diego Aquino
+ [Fast scroll](https://addons.mozilla.org/en-US/firefox/addon/fast-scroll/) by ddhillon



## Browser Requirements

+ Chrome 114 / Microsoft Edge 114
+ Firefox 114



## Usage

 Activation Modifier Key: Alt / option

| Action \\ Input \ Platform | Windows, Linux                                               | macOS                                                        |
| -------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| fast vertical scroll       | <kbd>Alt</kbd> + <code>WheelY</code>                         | <kbd>option</kbd> + <code>WheelY</code>                      |
| fast horizontal scroll     | <kbd>Alt</kbd> + <kbd>Shift</kbd> + <code>WheelY</code>, <kbd>Alt</kbd> + <code>WheelX</code> | <kbd>option</kbd> + <kbd>Shift</kbd> + <code>WheelY</code>, <kbd>option</kbd> + <code>WheelX</code> |
| fast zoom                  | <kbd>Alt</kbd> + <kbd>Ctrl</kbd> + <code>WheelY</code>       | <kbd>option</kbd> + <kbd>command</kbd> + <code>WheelY</code> |



## License

MIT &copy; 2023 Fuwei Chin

Icons used in this project are from open source project [Twemoji](https://twemoji.twitter.com/) which is licensed under [CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/)


CHANGELOG.md
# Changelog of Fast Scroll



## [0.2.0] - 2023-06-29

### Added

- Support Mac keyboard <kbd>option</kbd>, <kbd>command</kbd> as an equivalent of PC keyboard <kbd>Alt</kbd>, <kbd>Ctrl</kbd> for macOS
- Support instant scroll behavior
- Added scroll-by-page delta mode
- Added extension options page
- Added popup page to show extension usage

### Fixed

- Fixed scrollbar staggering on overscroll

### Changed

- Listen wheel event only when modifier key Alt/option pressed
- Respect web developers' `event.preventDefault()` to prevent wheel event from scrolling
- Ignore web developers' script dispatching wheel events on scrolling element.



## [0.1.0] - 2023-06-26

### Added

- Added fast vertical scroll, fast horizontal scroll
- Added fast zoom
- Support overlay scrollbars
- Support smooth scroll behavior


options.js
const $ = (s, c = document) => c.querySelector(s);
const $$ = (s, c = document) => Array.prototype.slice.call(c.querySelectorAll(s));
async function main() {
  let options = {fastScrollFactor: 3, fastScrollMode: 1};
  let factorInput = $('#fastScrollFactor');
  let modeInputs = $$('input[name="fastScrollMode"]');

  chrome.storage.sync.get(options, (result) => {
    options = result;
    factorInput.defaultValue = factorInput.value = options.fastScrollFactor;
    let checkedValue = '' + options.fastScrollMode;
    console.log('fastScrollMode', checkedValue);
    modeInputs.forEach((input) => {
      input.checked = input.value === checkedValue;
    });
  });

  modeInputs.forEach(function(input) { input.addEventListener('click', this); }, (e) => {
    let input = e.target;
    let mode = parseInt(input.value);
    console.log('fastScrollMode changed', mode);
    globalThis.getOptions().fastScrollMode = options.fastScrollMode = mode;
  });
  factorInput.addEventListener('change', (e) => {
    let input = e.target;
    if (!input.checkValidity()) {
      input.value = input.defaultValue;
    }
    let fastScrollFactor = parseInt(input.value);
    globalThis.getOptions().fastScrollFactor = options.fastScrollFactor = fastScrollFactor;
  });

  $('#save').addEventListener('click', (e) => {
    let button = e.target;
    button.disabled = true;
    chrome.storage.sync.set(options, () => {
      factorInput.defaultValue = options.fastScrollFactor;
      button.disabled = false;
    });
  });
}
queueMicrotask(main);


background.js
const env = {
  isFirefox: 'fileName' in new Error()
};

/* zoom */
let zoomValues = env.isFirefox ? [0.3, 0.5, 1, 2, 3, 4, 5] : [0.25, 0.333333, 0.5, 1, 2, 3, 4, 5];
let minZoomValue = Math.min(...zoomValues);
let maxZoomValue = Math.max(...zoomValues);

function nextZoom(current) {
  if (current < minZoomValue) {
    return minZoomValue;
  }
  for (let i = 0; i < zoomValues.length; i += 1) {
    let v = zoomValues[i];
    if (v <= current) {
      continue;
    }
    return v;
  }
  return maxZoomValue;
}
function prevZoom(current) {
  if (current > maxZoomValue) {
    return maxZoomValue;
  }
  for (let i = zoomValues.length - 1; i > -1; i -= 1) {
    let v = zoomValues[i];
    if (v >= current) {
      continue;
    }
    return v;
  }
  return minZoomValue;
}
async function zoomIn() {
  let [tab] = await chrome.tabs.query({active: true, lastFocusedWindow: true});
  if (!tab) {
    return;
  }
  let currentZoom = await chrome.tabs.getZoom(tab.id);
  currentZoom = +currentZoom.toFixed(6);
  if (currentZoom < maxZoomValue) {
    let zoomToBe = nextZoom(currentZoom);
    await chrome.tabs.setZoom(tab.id, zoomToBe);
  }
}
async function zoomOut() {
  let [tab] = await chrome.tabs.query({active: true, lastFocusedWindow: true});
  if (!tab) {
    return;
  }
  let currentZoom = await chrome.tabs.getZoom(tab.id);
  currentZoom = +currentZoom.toFixed(6);
  if (currentZoom > minZoomValue) {
    let zoomToBe = prevZoom(currentZoom);
    chrome.tabs.setZoom(tab.id, zoomToBe);
  }
}

/* scroll */
let scrollBehavior = 'auto';
function setScrollBehavior(behavior) {
  scrollBehavior = behavior;
}
function getScrollBehavior() {
  return scrollBehavior;
}

/* main */
let methods = {
  zoomIn,
  zoomOut,
  setScrollBehavior,
  getScrollBehavior,
};
chrome.runtime.onMessage.addListener(function(data, sender, sendResponse) {
  let method = methods[data.action];
  if (typeof method === 'function') {
    let returnValue;
    try {
      returnValue = method.apply(methods, 'value' in data ? [data.value] : []);
    } catch (e) {
      let reason = {name: e.name, message: e.message, stack: e.stack};
      sendResponse({status: 'rejected', reason});
      return;
    }
    if (method.constructor.name === 'AsyncFunction' ||
      returnValue && typeof returnValue.then === 'function') { // check if returnValue is thenable
      let responseId = crypto.randomUUID();
      sendResponse({status: 'pending', responseId});
      Promise.resolve(returnValue).then((value) => {
        chrome.tabs.sendMessage(sender.tab.id, {status: 'fulfilled', value, responseId}, {frameId: sender.frameId});
      }, (e) => {
        let reason = {name: e.name, message: e.message, stack: e.stack};
        chrome.tabs.sendMessage(sender.tab.id, {status: 'rejected', reason, responseId}, {frameId: sender.frameId});
      });
    } else {
      sendResponse({status: 'fulfilled', value: returnValue});
    }
  }
});


content.js
const env = {
  isFirefox: 'fileName' in new Error(),
  isMacOS: navigator.userAgentData?.platform === 'macOS' || navigator.platform.startsWith('Mac'),
};
const config = {
  lineHeight: env.isFirefox ? 100 / 6 : 100 / 3, // pre-knowledge of Windows platform, may vary on other platforms
  scrollBehavior: 'smooth', // to be detected, and cached to background.js
};
let options = { // to be synchronized from storage.sync
  fastScrollFactor: 3,
  fastScrollMode: 1, // 1: scroll by pixels, 2: scroll by pages
};
globalThis.getOptions = () => {
  return options;
};
let taskQueue = [];
let taskHandle = 0;

/* event handlers */
let yScrollingParentMap = new WeakMap();
let xScrollingParentMap = new WeakMap();
let scrollingElement = document.scrollingElement;
let isSmoothScrollWheel = false;

let wheelHandler = (e) => {
  let altKey = e.altKey;
  let ctrlKey = env.isMacOS ? e.metaKey : e.ctrlKey;
  if (altKey && !ctrlKey) {
    if (e.defaultPrevented || !e.isTrusted) // respect `event.preventDefault()`, ignore script-triggered event
      return;
    if (e.deltaMode === 2 || isSmoothScrollWheel) // skip if 'scroll by pages' detected, 'smooth scroll wheel' detected
      return;
    let {isFirefox} = env;
    if (isFirefox)
      e.preventDefault(); // prevent Alt + WheelY and Alt + WheelX from back-forward navigation
    let topOffset = 0;
    let leftOffset = 0;
    let axis;
    let {fastScrollFactor: factor, fastScrollMode} = options;
    if (e.deltaY) {
      if (e.deltaMode === 0 && Math.abs(e.deltaY) < 33) { // skip if 'smooth scroll wheel' detected
        isSmoothScrollWheel = true;
        return;
      }
      if (e.shiftKey) { // assign Shift + Alt + WheelY to fast horizontal scroll
        leftOffset = e.deltaMode === 0 ? e.deltaY : Math.round(e.deltaY * config.lineHeight);
        axis = 'x';
      } else { // assign Alt + WheelY to fast vertical scroll
        topOffset = e.deltaMode === 0 ? e.deltaY : Math.round(e.deltaY * config.lineHeight);
        axis = 'y';
      }
    } else if (e.deltaX) { // assign Alt + WheelX to fast horizontal scroll
      if (e.deltaMode === 0 && Math.abs(e.deltaX) < 33) { // skip if 'smooth scroll wheel' detected
        isSmoothScrollWheel = true;
        return;
      }
      leftOffset = e.deltaMode === 0 ? e.deltaX : Math.round(e.deltaX * config.lineHeight);
      axis = 'x';
    } else {
      return;
    }
    let scrollable;
    if (axis === 'y') {
      scrollable = yScrollingParentMap.get(e.target);
      if (!scrollable) {
        scrollable = getScrollableParent(e.target, axis);
        yScrollingParentMap.set(e.target, scrollable);
      }
      if (!canScrollYBy(scrollable, topOffset))
        return; // detected overscroll-y
    } else {
      scrollable = xScrollingParentMap.get(e.target);
      if (!scrollable) {
        scrollable = getScrollableParent(e.target, axis);
        xScrollingParentMap.set(e.target, scrollable);
      }
      if (!canScrollXBy(scrollable, leftOffset))
        return; // detected overscroll-x
    }
    if (scrollable !== scrollingElement) {
      if (isFirefox) {
        scrollable.scrollBy({
          left: leftOffset,
          top: topOffset,
          behavior: 'auto',
        }); // scroll immediately since default behavior was prevented
      }
      return; // only support fast scroll on document.scrollingElement
    }
    if (fastScrollMode === 2) {
      e.preventDefault();
      if (axis === 'y') {
        let sign = Math.sign(e.deltaY);
        scrollingElement.scrollBy(0, sign * scrollingElement.clientHeight);
      } else {
        let sign = (e.shiftKey && e.deltaY) ? Math.sign(e.deltaY) : Math.sign(e.deltaX);
        scrollingElement.scrollBy(sign * scrollingElement.clientWidth, 0);
      }
      return;
    }
    if (config.scrollBehavior === 'smooth') {
      let task = {left: leftOffset, top: topOffset, behavior: 'instant'};
      if (!isFirefox) {
        if (taskHandle !== 0) {
          e.preventDefault();
        } else {
          factor -= 1; // reduce scroll factor by 1 since default scroll allowed
        }
      }
      for (let i = 0; i < factor; i++) {
        taskQueue.push(task);
      }
    } else {
      if (!isFirefox)
        e.preventDefault();
      if (axis === 'y') {
        scrollingElement.scrollTo(0, scrollingElement.scrollTop + topOffset * factor);
      } else {
        scrollingElement.scrollTo(scrollingElement.scrollLeft + leftOffset * factor, 0);
      }
    }
  } else if (altKey && ctrlKey) {
    if (e.deltaY) {
      e.preventDefault(); // prevent Alt + Ctrl + WheelY from zooming
      let action =  e.deltaY < 0 ? 'zoomIn' : 'zoomOut';
      chrome.runtime.sendMessage({action}).then(handleDeferredResponse).then((_) => { // assign Alt + Ctrl + WheelY to fast zoom
        // NOOP
      });
    } else if (e.deltaX) {
      if (env.isFirefox)
        e.preventDefault(); // prevent  Alt + Ctrl + WheelX from horizontal scroll
    }
  }
};
let doScroll = () => {
  if (taskQueue.length > 0) {
    let task = taskQueue.pop();
    let {left, top} = task;
    if (top) {
      if (canScrollYBy(scrollingElement, top))
        scrollingElement.scrollBy(left, top);
    } else {
      if (canScrollXBy(scrollingElement, left))
        scrollingElement.scrollBy(left, top);
    }
    taskHandle = requestAnimationFrame(doScroll);
  } else {
    taskHandle = 0;
  }
};
let scrollHandler = (e) => {
  if (taskHandle === 0 && taskQueue.length > 0) {
    taskHandle = requestAnimationFrame(doScroll);
  }
};
let scrollendHandler = (e) => {
  if (taskHandle !== 0) {
    e.stopImmediatePropagation();
  }
  isSmoothScrollWheel = false;
};

/* messaging with background.js */
let deferred = {};
function handleDeferredResponse(data) {
  if (!data.responseId)
    throw new TypeError('No responseId');
  if (data.status !== 'pending')
    throw new TypeError('Not a pending response');
  return new Promise((resolve, reject) => {
    deferred[data.responseId] = {resolve, reject};
  });
}
function handleImmediateResponse(data) {
  if (data.status === 'fulfilled') {
    return data.value;
  } else {
    throw data.reason;
  }
}
chrome.runtime.onMessage.addListener((data) => {
  let {responseId} = data;
  if (responseId) {
    let obj = deferred[responseId];
    if (obj) {
      if (data.status === 'fulfilled') {
        obj.resolve(data.value);
      } else {
        obj.reject(data.reason);
      }
      delete deferred[responseId];
    }
  }
});

/* Fast Scroll API */
let fastScroll = false;
let requestFastScroll = () => {
  scrollingElement.addEventListener('wheel', wheelHandler, {passive: false});
  document.addEventListener('scroll', scrollHandler);
  document.addEventListener('scrollend', scrollendHandler);
  fastScroll = true;
};
let exitFastScroll = () => {
  scrollingElement.removeEventListener('wheel', wheelHandler, {passive: false});
  document.removeEventListener('scroll', scrollHandler);
  document.removeEventListener('scrollend', scrollendHandler);
  if (taskQueue.length > 0) {
    taskQueue.length = 0;
  }
  fastScroll = false;
};

/* main */
queueMicrotask(async function main() {
  document.body.addEventListener('keydown', (e) => {
    if (e.key === 'Alt') {
      if (!fastScroll)
        requestFastScroll();
    }
  });
  document.body.addEventListener('keyup', (e) => {
    if (e.key === 'Alt') {
      e.preventDefault(); // prevent menu button from focusing (Chrome), prevent menu bar from popup (Firefox)
      if (!e.getModifierState('Alt')) {
        if (fastScroll)
          exitFastScroll();
      }
    }
  });
  document.addEventListener('visibilitychange', () => { // in case of press of Alt + Tab, which keyup event of 'Tab' may not be triggered
    if (document.visibilityState === 'hidden') {
      if (fastScroll)
        exitFastScroll();
    }
  });
  chrome.storage.sync.get(options, (result) => {
    options = result;
  });
  chrome.storage.sync.onChanged.addListener((changes) => {
    let key = 'fastScrollFactor';
    if (key in changes) {
      options[key] = changes[key].newValue || 3;
    }
  });
  let behavior = await chrome.runtime.sendMessage({action: 'getScrollBehavior'}).then(handleImmediateResponse);
  if (behavior === 'auto' && document.visibilityState === 'visible') {
    let smooth = await detectSmoothScroll();
    behavior = smooth ? 'smooth' : 'instant';
    chrome.runtime.sendMessage({action: 'setScrollBehavior', value: behavior}).then(handleImmediateResponse);
  }
  config.scrollBehavior = behavior;
});

/* scroll utilities */
let rScrollable = ['scroll', 'auto'];
let isYScrollable = (el) => {
  return el.scrollHeight > el.clientHeight && rScrollable.includes(window.getComputedStyle(el).overflowY);
};
let isXScrollable = (el) => {
  return el.scrollWidth > el.clientWidth && rScrollable.includes(window.getComputedStyle(el).overflowX);
};
let getScrollableParent = function (el, axis) {
  if (axis === 'x') {
    for (let parent = el; parent !== null && parent !== scrollingElement; parent = parent.parentNode) {
      if (isXScrollable(parent)) {
        return parent;
      }
    }
  } else {
    for (let parent = el; parent !== null && parent !== scrollingElement; parent = parent.parentNode) {
      if (isYScrollable(parent)) {
        return parent;
      }
    }
  }
  return scrollingElement;
};
let canScrollYBy = (el, y) => y < 0 ? el.scrollTop > 0 : el.scrollHeight - el.clientHeight - el.scrollTop > 0;
let canScrollXBy = (el, x) => x < 0 ? el.scrollLeft > 0 : el.scrollWidth - el.clientWidth - el.scrollLeft > 0;
let detectSmoothScroll = () => {
  return new Promise((resolve) => {
    let resolved = false;
    let template = document.createElement('template');
    template.innerHTML = `<div style="position:absolute;left:-999px;top:-999px;width:360px;height:640px;overflow:auto;scroll-behavior:smooth;">
<div style="height:960px;"></div>
</div>`;
    let container = template.content.cloneNode(true).firstChild;
    let scrollCount = 0;
    let handler = (e) => { scrollCount++; };
    container.addEventListener('scroll', handler);
    container.addEventListener('scrollend', (e) => {
      if (resolved)
        return;
      resolved = true;
      container.removeEventListener('scroll', handler);
      container.remove();
      resolve(scrollCount > 1);
    }, {once: true});
    document.body.appendChild(container);
    queueMicrotask(() => {
      container.scrollBy({left: 0, top: container.scrollHeight - container.clientHeight, behavior: 'smooth'});
      setTimeout(() => {
        if (resolved)
          return;
        resolved = true;
        resolve(scrollCount > 1);
        container.remove();
      }, 400);
    });
  });
};
